File: app/src/main/java/com/chris/eban/common/EBanViewModelProviderFactory.java
package com.chris.eban.common;

import com.chris.eban.domain.usecase.EventListQuery;
import com.chris.eban.domain.usecase.EventSaveDelete;
import com.chris.eban.domain.usecase.EventSaveInsert;
import com.chris.eban.domain.usecase.EventSaveUpdate;
import com.chris.eban.presenter.event.EventDetailViewModel;
import com.chris.eban.presenter.event.EventListViewModel;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class EBanViewModelProviderFactory implements ViewModelProvider.Factory {

    private EventListQuery eventListQuery;
    private EventSaveInsert eventSaveInsert;
    private EventSaveUpdate eventSaveUpdate;
    private EventSaveDelete eventSaveDelete;

    public EBanViewModelProviderFactory
            (
                    EventListQuery eventListQuery,
                    EventSaveInsert eventSaveInsert,
                    EventSaveUpdate eventSaveUpdate,
                    EventSaveDelete eventSaveDelete
            ) {
        this.eventListQuery = eventListQuery;
        this.eventSaveInsert = eventSaveInsert;
        this.eventSaveUpdate = eventSaveUpdate;
        this.eventSaveDelete = eventSaveDelete;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(EventListViewModel.class)) {
            return (T) new EventListViewModel(eventListQuery, eventSaveDelete);
        } else if (modelClass.isAssignableFrom(EventDetailViewModel.class)) {
            return (T) new EventDetailViewModel(eventSaveInsert, eventSaveUpdate, eventSaveDelete);
        }
        throw new IllegalArgumentException("not support the class:" + modelClass);
    }
}

File: app/src/main/java/com/chris/eban/common/EBanViewModelProviderFactory.kt
package com.chris.eban.common

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.chris.eban.domain.usecase.*
import com.chris.eban.presenter.event.EventDetailViewModel
import com.chris.eban.presenter.event.EventListViewModel

class EBanViewModelProviderFactory(
        private val eventListQuery: EventListQuery,

        private val eventItemQuery: EventItemQuery,

        private val eventSaveInsert: EventSaveInsert,
        private val eventSaveUpdate: EventSaveUpdate,
        private val eventSaveDelete: EventSaveDelete
) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(EventListViewModel::class.java)) {
            return EventListViewModel(eventListQuery, eventSaveDelete) as T
        } else if (modelClass.isAssignableFrom(EventDetailViewModel::class.java)) {
            return EventDetailViewModel(eventItemQuery, eventSaveInsert, eventSaveUpdate, eventSaveDelete) as T
        }
        throw IllegalArgumentException("not support the class:$modelClass")
    }
}
